extpreimage/invoice.go

```diff
@@ -0,0 +1,124 @@
+package extpreimage
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// Invoice is a simplified view of a channeldb.Invoice that includes
+// the information needed to lookup an external preimage. We do
+// this to move as much logic out out of the main LND packages as
+// possible.
+type Invoice struct {
+	// ExternalPreimage indicates if the preimage for this hash is
+	// stored externally and must be retrieved.
+	ExternalPreimage bool
+
+	// PaymentHash is the hash that locks the HTLC for this payment.
+	PaymentHash [sha256.Size]byte
+
+	// PaymentPreimage is the preimage which is to be revealed in the
+	// occasion that an HTLC paying to the hash of this preimage is
+	// extended.
+	PaymentPreimage [32]byte
+
+	// Value is the expected amount of milli-satoshis to be paid to an HTLC
+	// which can be satisfied by the above preimage.
+	Value lnwire.MilliSatoshi
+
+	// Settled indicates if this particular contract term has been fully
+	// settled by the payer.
+	Settled bool
+}
+
+// InvoiceRegistry is a registry for storing invoices. It mirrors
+// the structure of main.InvoiceRegistry.
+type InvoiceRegistry interface {
+	AddInvoicePreimage(chainhash.Hash, [32]byte) error
+}
+
+// GetPaymentHash retrieves the payment hash for a given invoice,
+// either by calculating it from the preimage, or using the given
+// hash for invoices with external preimages.
+func (i *Invoice) GetPaymentHash() ([32]byte, error) {
+	var zeroHash [32]byte
+	var paymentHash [32]byte
+	var zeroPreimage [32]byte
+
+	if i.ExternalPreimage {
+		if bytes.Equal(i.PaymentHash[:], zeroHash[:]) {
+			return zeroHash, fmt.Errorf("Invoices with ExternalPreimage must " +
+				"have a locally defined PaymentHash.")
+		}
+
+		// For external preimages, we rely on a provided hash
+		paymentHash = i.PaymentHash
+	} else {
+		if bytes.Equal(i.PaymentPreimage[:], zeroPreimage[:]) {
+			return zeroHash, fmt.Errorf("Invoices must have a preimage or" +
+				"use ExternalPreimages")
+		}
+
+		// For local preimages, we calculate the hash ourselves
+		paymentHash = sha256.Sum256(i.PaymentPreimage[:])
+	}
+
+	return paymentHash, nil
+}
+
+// GetPaymentPreimage retrieves the preimage for a given invoice,
+// either by pulling it directly from the invoice, or by retrieving
+// it from the external preimage service if it is an external preimage
+// invoice.
+func (i *Invoice) GetPaymentPreimage(timeLock uint32, currentHeight uint32,
+	extpreimageClient Client, registry InvoiceRegistry) ([32]byte, error, error) {
+	var zeroPreimage [32]byte
+
+	switch {
+	// if there is a local preimage available, we should use it to settle the
+	// invoice
+	case !bytes.Equal(i.PaymentPreimage[:], zeroPreimage[:]):
```

Comment:
My initial reaction to adding the `switch` statement inside of `GetPaymentPreimage` is that we should leave the original code in `link.go` and detect if there is an external preimage... and then call out to this method.

**link.go**
```
switch {
        ....
	case !bytes.Equal(invoice.Terms.PaymentPreimage[:], zeroPreimage[:]):
		preimage = invoice.Terms.PaymentPreimage
	}
	// if this is an invoice with an external preimage, we should retrieve it.
	case invoice.Terms.ExternalPreimage:
                // make the call to extpreimage.GetPyamentPreimage
```

Reasoning for this is we dont want the `extpreimage` service to be responsible for ALL preimages (grabbing the preimage for all invoices), instead we would only want it in the case of identifying that there is an external preimage.

If we wanted any shared logic to determine if an Invoice is external or not, we should add those properties to `Invoices.go` itself.

We still have a problem where every time we would need a preimage, we would have conditional statements.

Anyways, with that said. What are your thoughts on the stuff above?

Reply:
In general, I agree. However, I moved the logic here so as to keep our diff with the upstream small. Everything in this file really belongs in invoices.go, but I put it here to make maintenance easier.

Reply:
@treygriffith Moving the ext preimages to an `invoice_preimage` module is a great change. From our talk, we decided against having the extpreimage package hold the responsibility of ALL invoice preimages. We've created another module to handle this which still allows maintenance to be easier (its a new file).